util.cpp

```diff
@@ -0,0 +1,55 @@
+#include <stdexcept>
+#include <boost/multiprecision/cpp_int.hpp>
+#include "include/util.h"
+
+using uint128_t = boost::multiprecision::uint128_t;
+
+const ull ULLMAX = std::numeric_limits<ull>::max();
+
+unsigned long long int double_to_ull(double d, double epsilon) {
+  return (unsigned long long) (d + epsilon);
+}
+
+ull nondirectional_non_self_edge_pairing_fn(ull i, ull j) {
+  // swap i,j if necessary
+  if (i > j) {
```

Comment:
use std::swap - it should be as fast or faster than this

Reply:
I don’t see how it’s faster: std::swap uses copy-construction, according to `cppreference`
```
template <class T> void swap ( T& a, T& b ) {   T c(a); a=b; b=c; }
```

Is this an empirical result you’ve seen in practice?

Reply:
Even if cppreference mentions that as a reference implementation (I can't find it), almost no library would ever implement it that way, and even if they did, modern compilers would be able to optimize the above significantly better than the xor trick. Also, that doesn't necessarily invoke copy construction - it's likely move construction at best.

See https://godbolt.org/z/54esx8 for a direct comparison in an arbitrary compiler - you can see std::swap generates fewer instructions with any optimization enabled.

Reply:
Perhaps even more notably, with NO optimization the reference implementation you give above is actually better:
https://godbolt.org/z/3o5Mnd
And with optimization it reduces to std;:swap

Reply:
interesting, I didn't know that. I'll change that